### PR TITLE
[AddressLowering] Handle indirectly yielded values used outside of coroutine range.

### PR DESCRIPTION
**Pull request by nate-chandler**
The `begin_apply` instruction introduces storage which is only valid until the coroutine is ended via `end_apply`/`abort_apply`.  Previously, AddressLowering assumed the memory was valid "as long as it needed to be": either as an argument to the function being lowered (so valid for the whole lifetime) or as a new `alloc_stack` (whose lifetime is determined by AddressLowering itself).  The result was that the storage made available by the `begin_apply` was used after the `end_apply`/`abort_apply` when there were uses of [a copy of] [a projection of] the yield after the `end_apply`/`abort_apply`.

Here, the behavior is fixed.  Now, the storage is used "as soon as possible".  There are two cases:

(1) If an indirect value's ownership is transferred into the caller (i.e. its convention is `@in` or `@in_constant`), the value is `copy_addr [take]`n into caller-local storage when lowering the `begin_apply`.

(2) If an indirect value's ownership is only lent to the caller (i.e. its convention is `@in_guaranteed`), no equivalent operation could be done: there's no `copy_addr [borrow]`; on the other hand, there's no need to do it--uses of the value must have occurred within the coroutine's range.  Instead, it's necessary to disable the store-copy optimization in this case: storage must be allocated for a copy of [a projection of] the yielded value.

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

File: lib/SIL/Utils/MemoryLocations.cpp [hunk 1]
```diff
@@ -173,6 +173,21 @@ void MemoryLocations::analyzeLocations(SILFunction *function) {
...
+      if (auto *BAI = dyn_cast<BeginApplyInst>(&I)) {
+        auto yieldedValues = BAI->getYieldedValues();
+        for (auto index : indices(yieldedValues)) {
+          auto fnType = BAI->getSubstCalleeType();
+          SILArgumentConvention argConv(
+              fnType->getYields()[index].getConvention());
+          if (argConv.isIndirectConvention()) {
+            // These are not SIL addresses under -enable-sil-opaque-values
+            if (!argConv.isInoutConvention() &&
+                !function->getConventions().useLoweredAddresses())
```
Comment (atrick) [inline lib/SIL/Utils/MemoryLocations.cpp:185]: You should never need to check `useLoweredAddresses()` when querying the argument convention. For code like this that cares about the SIL convention, you're supposed to use the SIL-level convention API, like `isSILIndirect`

**Comment by atrick** [comment]
What is `Indirect_In_Constant` equivalent to? `Indirect_In_Guaranteed` or `Indirect_In`? Our code-base seems to contradict itself. Can we just delete this convention in favor of one or the other?

**Comment by atrick** [comment]
@aschwaighofer do you have any idea what this convention means: `ParameterConvention::Indirect_In_Constant`? If not, I suggest removing it immediately. It's own documentation contradicts its usage, and it needs to be checked everywhere we handle conventions, so keeping it in the code results in a continuing proliferation of bugs.

File: lib/SIL/Utils/MemoryLocations.cpp (hunk 1 above)
Comment (nate-chandler) [reply lib/SIL/Utils/MemoryLocations.cpp]: Fixed.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test source compatibility

**Comment by atrick** [comment]
LGTM

**Pull request merged**